src/shadowbox/server/server_access_key.ts

```diff
@@ -56,28 +61,56 @@ function makeAccessKey(hostname: string, accessKeyJson: AccessKeyConfig): Access
       portNumber: accessKeyJson.port,
       encryptionMethod: accessKeyJson.encryptionMethod,
       password: accessKeyJson.password,
-    }
+    },
+    quota: accessKeyJson.quota,
+    isOverQuota: false
+  };
+}
+
+function makeAccessKeyJson(accessKey: AccessKey): AccessKeyJson {
+  return {
+    id: accessKey.id,
+    metricsId: accessKey.metricsId,
+    name: accessKey.name,
+    password: accessKey.proxyParams.password,
+    port: accessKey.proxyParams.portNumber,
+    encryptionMethod: accessKey.proxyParams.encryptionMethod,
+    quota: accessKey.quota
   };
 }
 
 // AccessKeyRepository that keeps its state in a config file and uses ShadowsocksServer
 // to start and stop per-access-key Shadowsocks instances.
 export class ServerAccessKeyRepository implements AccessKeyRepository {
-  // This is the max id + 1 among all access keys. Used to generate unique ids for new access keys.
   private NEW_USER_ENCRYPTION_METHOD = 'chacha20-ietf-poly1305';
+  private QUOTA_ENFORCEMENT_INTERVAL_MS = 3600000;  // 1h
+  private accessKeys: AccessKey[];
```

Comment:
Is there a particular reason why we don't use a Set\<AccessKey\>?

Reply:
One I can think of is indexed access being useful in tests.

Reply:
Also it looks like Sets of object types are pretty janky in Typescript since it uses identity to compare for equality.  So never mind on that one :)